crates/pessimistic-proof/src/global_state.rs

```diff
@@ -0,0 +1,67 @@
+use std::collections::BTreeMap;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    certificate::Certificate,
+    local_balance_tree::BalanceTree,
+    local_exit_tree::{hasher::Keccak256Hasher, LocalExitTree},
+    proof::{BalanceRoot, ExitRoot},
+    BridgeExit, NetworkId, ProofError,
+};
+
+/// Local state of one network.
+/// The AggLayer tracks the [`LocalNetworkState`] for all networks.
+/// Eventually, this state will be entirely tracked by the networks themselves.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LocalNetworkState {
+    /// Commitment to the [`BridgeExit`].
+    pub exit_tree: LocalExitTree<Keccak256Hasher>,
+    /// Commitment to the balance for each token.
+    pub balance_tree: BalanceTree,
+    /// Commitment to the imported [`BridgeExit`].
+    #[allow(dead_code)]
+    pub nullifier_tree: BTreeMap<NetworkId, Vec<BridgeExit>>,
+}
+
+impl LocalNetworkState {
+    /// Apply the [`Certificate`] on the current [`State`].
+    /// Returns the commitment on the resulting state if successful.
+    pub fn apply_certificate(
+        &mut self,
+        certificate: Certificate,
+    ) -> Result<(ExitRoot, BalanceRoot), ProofError> {
+        // Apply on Exit Tree
+        {
+            let computed_root = self.exit_tree.get_root();
```

Comment:
Could we not compute the exit root each time but having the `local_exit_root` as a value of `LocalNetworkState` ?

Reply:
This is the code that will be run by the prover, so we want to ensure that the initial state is properly pegged to a given initial root, so in this computation we do want to compute the initial root based on the initial local exit tree